File: src/persist_sqlcipher.py
# ============================================================
# persist_sqlcipher.py — SQLCipher adapter for persist
#
# Encrypted SQLite.  Same API as sqlite_adapter, plus PRAGMA key.
# Requires pysqlcipher3 or sqlcipher3:
#   pip install pysqlcipher3
#
# Usage:
#   from persist_sqlcipher import sqlcipher_adapter
#   persist(engine, sqlcipher_adapter("state.db", "secret"))
# ============================================================


def sqlcipher_adapter(path, key):
    """Create a SQLCipher persist adapter.

    path — file path (created if missing)
    key  — encryption passphrase
    """
    try:
        from pysqlcipher3 import dbapi2 as db_mod
    except ImportError:
        import sqlcipher3 as db_mod

    conn = db_mod.connect(path)
    conn.execute("PRAGMA key = ?", (key,))
    conn.execute("PRAGMA journal_mode=WAL")

    def _setup():
        conn.execute(
            "CREATE TABLE IF NOT EXISTS facts "
            "(term TEXT PRIMARY KEY, functor TEXT, arity INTEGER)"
        )
        conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_facts_pred ON facts(functor, arity)"
        )
        conn.commit()

    def _all(predicates=None):
        if predicates:
            rows = []
            for pred in predicates:
                parts = pred.split("/")
                rows.extend(conn.execute(
                    "SELECT term FROM facts WHERE functor = ? AND arity = ?",
                    (parts[0], int(parts[1]))
                ).fetchall())
            return [r[0] for r in rows]
        return [r[0] for r in conn.execute("SELECT term FROM facts")]

    return {
        "setup":  _setup,
        "insert": lambda key, functor=None, arity=None: conn.execute(
            "INSERT OR IGNORE INTO facts VALUES (?, ?, ?)", (key, functor, arity)
        ),
        "remove": lambda key: conn.execute(
            "DELETE FROM facts WHERE term = ?", (key,)
        ),
        "all":    _all,
        "commit": lambda: conn.commit(),
        "close":  lambda: conn.close(),
    }

File: src/persist-sqlcipher.js
// ============================================================
// persist-sqlcipher.js — SQLCipher adapter for persist
//
// Encrypted SQLite.  Same as sqliteAdapter + PRAGMA key.
// Requires better-sqlite3 compiled with SQLCipher support,
// or @journeyapps/sqlcipher.
//
// Usage:
//   persist(engine, sqlcipherAdapter(new Database('state.db'), 'secret'));
// ============================================================

function sqlcipherAdapter(db, key) {
  db.pragma("key = '" + key + "'");

  var cache = {};
  function stmt(sql) {
    if (!cache[sql]) cache[sql] = db.prepare(sql);
    return cache[sql];
  }

  return {
    setup: function() {
      db.exec(
        "CREATE TABLE IF NOT EXISTS facts " +
        "(term TEXT PRIMARY KEY, functor TEXT, arity INTEGER)"
      );
      db.exec(
        "CREATE INDEX IF NOT EXISTS idx_facts_pred ON facts(functor, arity)"
      );
    },
    insert: function(key, functor, arity) {
      stmt("INSERT OR IGNORE INTO facts VALUES (?, ?, ?)").run(key, functor, arity);
    },
    remove: function(key) {
      stmt("DELETE FROM facts WHERE term = ?").run(key);
    },
    all: function(predicates) {
      if (predicates) {
        var rows = [], keys = Object.keys(predicates);
        for (var i = 0; i < keys.length; i++) {
          var parts = keys[i].split("/");
          var matched = stmt("SELECT term FROM facts WHERE functor = ? AND arity = ?")
            .all(parts[0], parseInt(parts[1], 10));
          for (var j = 0; j < matched.length; j++) rows.push(matched[j].term);
        }
        return rows;
      }
      return stmt("SELECT term FROM facts").all().map(function(r) { return r.term; });
    },
    commit: function() {},
    close: function() { db.close(); }
  };
}

// ── Export (dual ESM/CJS) ───────────────────────────────────

if (typeof exports !== "undefined") {
  exports.sqlcipherAdapter = sqlcipherAdapter;
}
export { sqlcipherAdapter };
